action/binary.go

```
@@ -0,0 +1,185 @@
+		// copying from one secret to another secret is not supported
+		return fmt.Errorf("ambiquity detected. Either from or to must be a file")
+	case fsutil.IsFile(from) && !fsutil.IsFile(to):
+		// if the source is a file the destination must no to avoid ambiquities
+		// if necessary this can be resolved by using a absolute path for the file
+		// and a relative one for the secret
+		if !strings.HasSuffix(to, BinarySuffix) {
+			to += BinarySuffix
+		}
+		// copy from FS to store
+		buf, err := ioutil.ReadFile(from)
+		if err != nil {
+			return err
+		}
+		if err := s.Store.Set(to, []byte(base64.StdEncoding.EncodeToString(buf)), fmt.Sprintf("Copied data from %s to %s", from, to)); err != nil {
+			return err
+		}
+		if deleteSource {
+			if err := s.binaryValidate(buf, to); err != nil {
+				return err
+			}
+			return fsutil.Shred(from, 8)
+		}
+		return nil
+	case !fsutil.IsFile(from):
+		// if the source is no file we assume it's a secret and to is a filename
+		// (which may already exist or not)
+		if !strings.HasSuffix(from, BinarySuffix) {
+			from += BinarySuffix
+		}
+		// copy from store to FS
+		buf, err := s.binaryGet(from)
+		if err != nil {
+			return err
+		}
+		if err := ioutil.WriteFile(to, buf, 0600); err != nil {
+			return err
+		}
+		if deleteSource {
+			if err := s.binaryValidate(buf, from); err != nil {
+				return err
+			}
+			return s.Store.Delete(from)
+		}
+		return nil
+	default:
+		return fmt.Errorf("ambiquity detected. Unhandled case. Please report a bug")
+	}
+}
+
+func (s *Action) binaryValidate(buf []byte, name string) error {
+	h := sha256.New()
+	_, _ = h.Write(buf)
+	fileSum := fmt.Sprintf("%x", h.Sum(nil))
+
+	h.Reset()
+
+	var err error
+	buf, err = s.binaryGet(name)
+	if err != nil {
+		return err
+	}
+	_, _ = h.Write(buf)
+	storeSum := fmt.Sprintf("%x", h.Sum(nil))
+
+	if fileSum != storeSum {
+		return fmt.Errorf("Hashsum mismatch (file: %s, store: %s)", fileSum, storeSum)
```

if we plan to expose this function (via wrapper) to the UI, we should return a custom error type, so we can handle error in retrieving a store secret and a validation mismatch differently